File: leetcode/stack/2751.cpp
#include <algorithm>
#include <string>
#include <tuple>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> survivedRobotsHealths(vector<int>& positions, vector<int>& healths, string directions) {

        // size with vector<int>&
        int n = positions.size();

        vector<tuple<int, int, char>> arr;
        for (int i = 0; i < n; i++) {
            arr.push_back({positions[i], healths[i], directions[i]});
        }
        sort(arr.begin(), arr.end());

        // stack
        vector<tuple<int, int, char>> st;

        for (auto& [pos, hp, dir] : arr) {
            if (dir == 'R') {
                st.push_back({pos, hp, dir});
            }
            else {
                bool can_push = true;
                while (!st.empty() && get<2>(st.back()) == 'R') {
                    int& collision_hp = get<1>(st.back());
                    if (collision_hp < hp) {
                        hp--;
                        st.pop_back();
                    }
                    else if (collision_hp == hp) {
                        st.pop_back();
                        can_push = false;
                        break;
                    }
                    else if (collision_hp > hp) {
                        can_push = false;
                        collision_hp -= 1;
                        break;
                    }
                }

                if (can_push) {
                    st.push_back({pos, hp, dir});
                }
            }
        }
        
        unordered_map<int, int> mp;
        for (auto& [pos, hp, dir] : st) {
            mp[pos] = hp;
        }

        vector<int> res;
        for (int pos : positions) {
            if (mp.count(pos)) {
                res.push_back(mp[pos]);
            }
        }

        return res;
    }
};

File: leetcode/stack/2751.py
from typing import List


class Solution:
    def survivedRobotsHealths(self, positions: List[int], healths: List[int], directions: str) -> List[int]:
        stack = []
        arr = sorted(zip(positions, healths, directions))
        for pos, hp, direction in arr:
            if direction == "R":
                stack.append([pos, hp, direction])

            if direction == "L":
                can_push = True

                while stack and stack[-1][2] == "R":
                    if stack[-1][1] < hp:
                        stack.pop()
                        hp -= 1
                    elif stack[-1][1] == hp:
                        stack.pop()
                        can_push = False
                        break
                    elif stack[-1][1] > hp:
                        stack[-1][1] -= 1
                        can_push = False
                        break
                
                if can_push:
                    stack.append([pos, hp, direction])

        d = dict()
        for pos, hp, direction in stack:
            d[pos] = hp

        res = [d[pos] for pos in positions if pos in d]
        return res
